### PR TITLE
[BEAM-9919] Isolating xlang transforms by namespace

### PR DESCRIPTION
**Pull request by pskevin**
All components of the xlang transform that aren't uniquely identified by topological ordering are now isolated into their respective namespaces.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels
Python tests

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pskevin** [comment]
R: @lostluck

**Comment by pskevin** [comment]
CC: @chamikaramj

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	coder, exists := c.Coders[cid]
+	if !exists {
+		panic(errors.Errorf("attempted to add namespace to missing coder id: %v not in %v", cid, c.Coders))
+	}
+
+	// Updating ComponentCoderIDs of Coder
+	if coder.GetComponentCoderIds() != nil {
+		updatedComponentCoderIDs := coder.ComponentCoderIds // Pass by value
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/namespace.go:37]: Slices are always passed by reference, so this is overwriting the coder lists in the extracted coder. You would need to create a n new slice, and append values to that and  *then* overwrite the existing one in coder.

updatedComponentCoderIDs := append(nil, coder.ComponentCoderIds...)
That should create a new slice with the same elements.

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		for i, ccid := range coder.ComponentCoderIds {
+			updatedComponentCoderIDs[i] = AddCoderID(c, idMap, ccid, newID)
+			coder.ComponentCoderIds = updatedComponentCoderIDs
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/namespace.go:40]: Should this assignment be outside the loop?

This is saying after each element in coder.ComponentCoderIds, reset the ComponentCoderIds to this new updated version.

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+	updateEnvironmentID := func(eid string) string {
+		return AddEnvironmentID(c, idMap, eid, newID)
+	}
+
+	// Update Environment ID of PTransform
+	if t.EnvironmentId != "" {
+		t.EnvironmentId = updateEnvironmentID(t.EnvironmentId)
+	}
+	fmt.Println(t)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/namespace.go:124]: leftover Debug printout

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		return AddCoderID(c, idMap, cid, newID)
+	}
+
+	updateWindowingStrategyID := func(wid string) string {
+		return AddWindowingStrategyID(c, idMap, wid, newID)
+	}
+
+	updateEnvironmentID := func(eid string) string {
+		return AddEnvironmentID(c, idMap, eid, newID)
+	}
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/xlangx/namespace.go:118]: Remove these helpers, and call the functions directly in the loop below please. They're only used once each, and you do not modify any of their parameters, making the closures less useful, and harder to read.

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go (hunk 1 above)
Comment (pskevin) [reply sdks/go/pkg/beam/core/runtime/xlangx/namespace.go]: Ack. Updated!

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go (hunk 2 above)
Comment (pskevin) [reply sdks/go/pkg/beam/core/runtime/xlangx/namespace.go]: Ack. Updated!

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go (hunk 3 above)
Comment (pskevin) [reply sdks/go/pkg/beam/core/runtime/xlangx/namespace.go]: Sorry about that. Removed it.

File: sdks/go/pkg/beam/core/runtime/xlangx/namespace.go (hunk 4 above)
Comment (pskevin) [reply sdks/go/pkg/beam/core/runtime/xlangx/namespace.go]: Makes sense. Removed it!

**Pull request merged**